File: Server/netty-server/src/main/java/yayuki/heartrate/server/handler/StateHandler.java
package yayuki.heartrate.server.handler;

import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import yayuki.heartrate.net.protobuf.PatientProto;
import yayuki.heartrate.server.repository.ChannelRepository;

public class StateHandler extends SimpleChannelInboundHandler<PatientProto.State> {
    private int carerID;

    public StateHandler(int carerID) {
        this.carerID = carerID;
    }

    @Override
    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        super.channelRegistered(ctx);
    }

    @Override
    protected void channelRead0(ChannelHandlerContext ctx, PatientProto.State msg) throws Exception {
        Channel carerChannel = ChannelRepository.INSTANCE.getCarerChannel(carerID);
        if (carerChannel == null) return;
        carerChannel.writeAndFlush(msg);
    }
}

File: Android/netty-carer/src/main/java/yayuki/heartrate/client/carer/handler/StateHandler.kt
package yayuki.heartrate.client.carer.handler

import io.netty.channel.ChannelHandler
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.SimpleChannelInboundHandler
import kotlinx.coroutines.channels.BufferOverflow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.runBlocking
import yayuki.heartrate.carer.data.PatientStateEvent
import yayuki.heartrate.net.protobuf.PatientProto

@ChannelHandler.Sharable
class StateHandler : SimpleChannelInboundHandler<PatientProto.State>() {
    val stateFlow = MutableSharedFlow<PatientStateEvent>(10, 5, BufferOverflow.DROP_OLDEST)
    override fun channelRead0(ctx: ChannelHandlerContext?, msg: PatientProto.State) = runBlocking {
        stateFlow.emit(msg.run {
            PatientStateEvent(id, heartRate, accuracy, latitude, longitude)
        })
    }
}